src/TSTransformer/nodes/statements/transformForOfStatement.ts

```diff
@@ -409,6 +411,17 @@ const buildGeneratorLoop: LoopBuilder = makeForLoopBuilder((state, initializer,
 });
 
 function getLoopBuilder(state: TransformState, node: ts.Node, type: ts.Type): LoopBuilder {
+	// Check for 'any' or 'unknown' types first to provide better error messages
+	if (isDefinitelyType(type, isAnyType(state))) {
+		DiagnosticService.addDiagnostic(errors.noAnyForOfIteration(node));
+		return () => luau.list.make();
+	}
```

Comment:
This approach at the top of the function means there are still some types that cause assertion errors ([for example](https://osyrisrblx.github.io/roblox-ts.com/playground/#code/MYGwhgzhAECC0G8CwAoAkMA9gOwgFwCcBXYPTAgCgEpFU00Azc6CrXPaMaTB6PACwCWEGsnRoAvnSnoA2gGUAngFsARphAA6QXgCmBMGQIBdarXEFdeIgWzn6abLoAeeM2IdpL124mgA3MBAiXQAuaABGaAkAbjp6CQAaeNjpVBkgA), iterating on `this` inside a class). It also introduces a small performance regression, because we have an extra typecheck being done for every loop. It would be better to make the diagnostic a bit more generic and have it replace the assertion at the end.